File: src/main/java/model/organization/event/AttributeEvent.java
package model.organization.event;

import static model.organization.validation.Checks.checkNotNull;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import model.organization.entity.Attribute;
import model.organization.id.UniqueId;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain AttributeEvent} class indicates that there is an update about an {@linkplain Attribute} entity.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class AttributeEvent extends AbstractEvent {

	private static final long serialVersionUID = -2314619272565535715L;
	private final UniqueId<Attribute> id;

	@Inject
	AttributeEvent(@NotNull @Assisted final Attribute attribute, @NotNull @Assisted final EventCategory category) {
		super(category);
		checkNotNull(attribute, "attribute");
		id = attribute.getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Attribute}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Attribute> getId() {
		return id;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/AttributeEvent.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.AttributeId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.model.event.AbstractEvent
import runtimemodels.chazm.model.event.EventType
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [AttributeEvent] class indicates that there is an update about an [Attribute] entity.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class AttributeEvent @Inject internal constructor(
    category: EventType,
    attribute: Attribute
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Attribute].
     *
     * @return a [UniqueId].
     */
    val id: AttributeId = attribute.id

    override fun equals(other: Any?): Boolean {
        if (other is AttributeEvent) {
            return super.equals(other) && id == other.id
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, id)

    override fun toString(): String = M.EVENT_WITH_1_ID[super.toString(), id]

    companion object {
        private const val serialVersionUID = -2314619272565535715L
    }

}
